spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java

```diff
@@ -1088,13 +1088,27 @@ private void logicalClose(ChannelProxy proxy) throws Exception {
 					}
 				}
 			}
-			// Allow for multiple close calls...
-			if (!this.channelList.contains(proxy)) {
-				if (logger.isTraceEnabled()) {
-					logger.trace("Returning cached Channel: " + this.target);
+			if (CachingConnectionFactory.this.active && this.publisherConfirms
+					&& proxy instanceof PublisherCallbackChannel) {
+				this.theConnection.channelsAwaitingAcks.add(proxy);
+				((PublisherCallbackChannel) proxy).setAfterAckCallback(this::returnToCache, proxy);
```

Comment:
This a bit awkward, but I guess we don't have choice since `PublisherCallbackChannelImpl` is not `ChannelProxy`  instance.
I mean we we might not be able just to use `this` there in the `processAck()`. Correct ?

Reply:
Yeah - `this` doesn't work (that's what I had initially) - we need a reference to the proxy. We can look at reworking in 2.1.

Reply:
... or we need to keep a dictionary for `PublisherCallbackChannel <-> ChannelProxy`  cross-referencing.